arxiv/metadata/metacheck.py

```diff
@@ -0,0 +1,687 @@
+SOURCE_TYPE = FieldName.SOURCE_TYPE
+COMMENTS = FieldName.COMMENTS
+REPORT_NUM = FieldName.REPORT_NUM
+JOURNAL_REF = FieldName.JOURNAL_REF
+DOI = FieldName.DOI
+
+OK = Disposition.OK
+WARN = Disposition.WARN
+HOLD = Disposition.HOLD
+
+############################################################
+# Some constants
+
+MIN_TITLE_LEN = 5
+MIN_AUTHORS_LEN = 5
+MIN_ABSTRACT_LEN = 5
+MIN_SOURCE_TYPE_LEN = 0         # Not checked
+# No MIN_COMMENTS_LEN
+MIN_REPORT_NUM_LEN = 4
+MIN_JOURNAL_REF_LEN = 5
+MIN_DOI_LEN = 10                # such as: 10.1234/5678 
+
+# Language identification is pretty fragile, particularly on short abstracts.
+MIN_CHARS_FOR_LID = 20
+
+############################################################
+
+def combine_dispositions(d1:Disposition, d2:Disposition) -> Disposition:
+    if d1 == HOLD or d2 == HOLD:
+        return HOLD
+    elif d1 == WARN or d2 == WARN:
+        return WARN
+    else:
+        return OK
+
+def contains_outside_math(s1:str, s2:str) -> bool:
+    """ Looks for s1, outside of TeX math
+    Not perfect: fails to find xyzzy in $math$ xyzzy $$more math$$.
+    """
+    if re.search(s1, s2, re.IGNORECASE) \
+       and not re.search("[$].*"+s1+".*[$]", s2, re.IGNORECASE):
+        return True
+    else:
+        return False
+       
+
+############################################################
+# This is the main function
+
+def check(metadata:Dict[FieldName,str]): 
+    result = {} 
+    for (k,v) in metadata.items():
+        if k == TITLE:
+            result[k] = check_title(v)
+        elif k == AUTHORS:
+            result[k] = check_authors(v)
+        elif k == ABSTRACT:
+            result[k] = check_abstract(v)
+        elif k == SOURCE_TYPE:
+            result[k] = check_source_type(v)
+        elif k == COMMENTS:
+            result[k] = check_comments(v)
+        elif k == REPORT_NUM:
+            result[k] = check_report_num(v)
+        elif k == JOURNAL_REF:
+            result[k] = check_journal_ref(v)
+        elif k == DOI:
+            result[k] = check_doi(v)
+        else:
+            raise UnknownMetadataException(f"Unknown metadata field {k}: {v}")
+    #
+    return result
+
+def looks_like_all_caps(s: str) -> bool:
+    """
+    Returns true if s appears to have all/excessive capitals. The
+    exact threshold is heuristic based on looking at submissions over
+    March 2010.
+    """
+    num_caps = s.upper()
+    num_caps = sum([ c.isupper() for c in s])
+    num_lower = sum([ c.islower() for c in s])
+    if num_caps > num_lower * 2 + 7:
+        return True
+    else:
+        return False
+
+
+def long_word_caps(s: str) -> bool:
+    """
+    Returns true (was: a list of long words) if s appears to have
+    two or more words which are not in our list
+    KNOWN_WORDS_IN_ALL_CAPS.
+
+    """
+    num_matches = 0
+    for word in s.split():
+        if len(word) >= 6 and \
+           word.upper() == word and \
+           word.lower() != word and \

```

Comment:
Seems redundant with line 116

Reply:
This is intentional, and not redundant.

word.upper() == word checks that all alpha letters in word are in Upper case, not lower case.
word.lower() != word checks that the word actually contains some alpha (and does not consist only of digits and punctuation).

Reply:
word.lower() != word will be True when there are upper case letters in word, which there must be if word.upper() == word. 
I still feel it is redundant and may not be doing what you want done.

Reply:
Put a different way: word.upper() == word checks that there are *no lowercase letters* in the word, while word.lower() != word checks that there are some uppercase letters in the word.

I'll make sure there are unit tests which demonstrate both parts of the test, so you can comment out one of the two lines and see which tests fail.